### ISSUE TITLE
Samples not stopping when they should

### ISSUE THREAD

**Issue by VAdaPEGA** [vladikcomper/MegaPCM#3]
I'm having this odd issue where for some samples, it continues playing past where it should stop, making them play a bit of the next sample by mistake.
Not sure if it's my set up or something wrong with my wav files.

Using asm68k, problems started after migration from MegaPCM 1.x to MegaPCM 2

https://github.com/user-attachments/assets/43bfeb0b-0342-419f-a9b2-33304dffb2ba

Here I play a couple of samples, some working ones and some with the problematic ends.
$A0 being especially bad
$9E being the most subtle, leaving an awful click at the end
(not recorded : $AA seems to have syncing problems, this was not the case on MegaPCM 1.x)

[SoundDriver_DACOnly.zip](https://github.com/user-attachments/files/17396268/SoundDriver_DACOnly.zip)
Here's what I have, if anything seems wrong about my setup, let me know.

EDIT : just tested it out with https://github.com/vladikcomper/s1disasm-megapcm2 , seems it is the samples and settings that are giving trouble.

**Comment by vladikcomper** [comment]
I linked this sample table to the debug version of Mega PCM 2 and it logged a few interesting things. I have a feeling some WAV files are corrupted (at least that's what it looks like).

Output of problematic samples for example ($A0):
```
KDEBUG MESSAGE: Sample: type=50, flags=00, pitch=A2, start=guimokick2, end=guimokick2_end
KDEBUG MESSAGE: Detected WAVE header
KDEBUG MESSAGE: Seeking data chunk: Skipping 00000010 bytes
KDEBUG MESSAGE: Checking data chunk at guimokick2+24
KDEBUG MESSAGE: WAVE data offsets: start=guimokick2+2C, end=guimosnare2+68A
```

The sample record specified `start=guimokick2, end=guimokick2_end`, but since it's a WAV file Mega PCM 2 went ahead to locate the exact boundaries of WAVE file's `data` block (so it won't attempt to "play" headers and metadata/tags). And this is where overflow comes from: `WAVE data offsets: start=guimokick2+2C, end=guimosnare2+68A` (`data` block pointers seem to overflow into the next sample).

Here's another one, similarly corrupted:
```
KDEBUG MESSAGE: Sample: type=50, flags=00, pitch=DF, start=guimosnare2, end=guimosnare2_end
KDEBUG MESSAGE: Detected WAVE header
KDEBUG MESSAGE: Seeking data chunk: Skipping 00000010 bytes
KDEBUG MESSAGE: Checking data chunk at guimosnare2+24
KDEBUG MESSAGE: WAVE data offsets: start=guimosnare2+2C, end=guimosnare3+92
```

Fill log of loaded samples for reference:
[output.log](https://github.com/user-attachments/files/17400342/output.log)

So something's up with these files. Were they abruptly truncated without fixing the header?

I need to inspect this more closely. I'll get back with the results soon.

**Comment by vladikcomper** [comment]
FFmpeg also doesn't like these samples:
```ini
$ ffprobe -i GuimO/kick2.wav -hide_banner                                                                                          
[wav @ 0x55886a35f900] Packet corrupt (stream = 0, dts = NOPTS).
[wav @ 0x55886a35f900] Estimating duration from bitrate, this may be inaccurate
Input #0, wav, from 'GuimO/kick2.wav':
  Duration: 00:00:00.39, bitrate: 67 kb/s
  Stream #0:0: Audio: pcm_u8 ([1][0][0][0] / 0x0001), 8272 Hz, 1 channels, u8, 66 kb/s
```

I wonder how this happened though. Were they edited by some old glitchy tool or truncated at some point?

**Comment by vladikcomper** [comment]
Full list of corrupted files according to `ffprobe`:

```ini
Scrsh2sy78.wav
[wav @ 0x563809214900] Packet corrupt (stream = 0, dts = NOPTS).

VCome_on.wav
[wav @ 0x55b457a89900] Packet corrupt (stream = 0, dts = NOPTS).

woodyyeah.wav
[wav @ 0x559a04b79900] Packet corrupt (stream = 0, dts = NOPTS).

GuimO/clap.wav
[wav @ 0x55cdd3159900] Packet corrupt (stream = 0, dts = NOPTS).

GuimO/kick2.wav
[wav @ 0x55d7bfc7b900] Packet corrupt (stream = 0, dts = NOPTS).

GuimO/Snare2.wav
[wav @ 0x55b9037b6900] Packet corrupt (stream = 0, dts = NOPTS).

GuimO/wood.wav
[wav @ 0x55ed957be900] Packet corrupt (stream = 0, dts = NOPTS).
```

**Comment by VAdaPEGA** [comment]
Ok this makes a whole lot of sense if that's how MegaPCM2 detects its borders.
On MegaPCM1 I would manually truncate the samples via hex editor rather than on an audio editor to save on some space, neglecting the header entirely.
So essentially, all I have to do is fix the headers on the wav files.

**Comment by vladikcomper** [comment]
@VAdaPEGA Here, I fixed it for you =P

[vadapega-wav-fixed.tar.gz](https://github.com/user-attachments/files/17402456/vadapega-wav-fixed.tar.gz)

I just used SOX utility on Linux to copy audiostreams as-is and it automatically fixed the header. I could've used FFMpeg as well, but it adds an extra 22 byte section in WAVE with codec metadata (totally harmless, just space wasted).

```sh
$ mkdir fixed
$ for i in *.wav ; do sox $i fixed/$i ; done
sox WARN wav: Premature EOF on .wav input file
sox WARN wav: Premature EOF on .wav input file
sox WARN wav: Premature EOF on .wav input file
sox WARN wav: Premature EOF on .wav input file
```

I didn't test it yet, but I'm sure it should play fine now.

And yeah, Mega PCM 2 is quite high-level now; so it reads actual WAVE headers and plays according to them (when uploading sample table to Z80 in `MegaPCM_LoadSampleTable`), no more cheap tricks of skipping bytes like in 1.x.

I'm glad to hear it was just abruptly truncated though; so I don't have to deal with potential broken WAVE editors and their edge cases.

**Comment by VAdaPEGA** [comment]
Whoa awesome, thanks for the help!

**Comment by vladikcomper** [comment]
@VAdaPEGA I just noticed you mentioned sample $AA had desync problems. I took a closer look and it turned out to be a legit bug! You triggered quite a rare edge case with unusually low sample rate (4 kHz).

Can you try this patched version? It should fix desync issues with extremely low-rate samples:

[megapcm-2.0.1.tar.gz](https://github.com/user-attachments/files/17425219/megapcm-2.0.1.tar.gz)

Here's a demo ROM I used to test this (sample $AA seems to play properly here):
[rom.zip](https://github.com/user-attachments/files/17425234/rom.zip)

**In case you wonder what the bug is:**
Mega PCM 2 always buffers samples at double the max rate (50416 Hz in case of PCM) so it can survive up to 33% of CPU time without ROM access by playing solely from a 256-byte "read ahead" buffer (in Z80 RAM). On the other hand, it plays previously buffered data in parallel (it's always buffer 2 samples @50kHz, play 1 sample at the desired rate). High enough rates keep the balance: the buffer is only half-full most of the time. But 4 kHz is so low, the buffer is always full and accepts new samples every other time.

This isn't an issue though: I always check if buffer is full in the playback loop. Except for one edge case: if "read ahead" buffer is full and Z80 was about to switch banks, it mistakenly buffered first 2 samples from the next bank without checking. This pushed buffer pointer beyond a threshold and caused overrun for one iteration... which would eat up and skip ~256 samples. And again, since it's 4 Khz, 256 samples become noticeable.

**Comment by VAdaPEGA** [comment]
I was gonna make a separate issue for the weird syncing problems but was too tired to write up about it, thanks for looking into it!
The patched version works well!
